### PR TITLE
Add package signing

### PR DESCRIPTION
**Pull request by martincostello**
Add Authenticode signing to the build process between publishing to GitHub Packages (effectively a staging area) and NuGet.org (being "production").

The signing job was adapted from [this example](https://github.com/dotnet/sign/blob/main/docs/gh-build-and-sign.yml) and extended to validate that the packages and their contents were correctly signed.

I tested the signature validation using this PR in a branch of one of my own libraries which I don't Authenticode sign: https://github.com/martincostello/xunit-logging/pull/438 ([validation failure](https://github.com/martincostello/xunit-logging/actions/runs/5223505072/jobs/9430504223)).

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -131,28 +138,106 @@ jobs:
...
+    - name: Install Sign CLI tool
+      run: dotnet tool install --tool-path . sign --version 0.9.0-beta.23127.3
+
+    - name: Azure CLI login
+      uses: azure/login@92a5484dfaf04ca78a94597f4f19fea633851fa2 # v1.4.7
+      with:
+        allow-no-subscriptions: true
+        client-id: ${{ secrets.AZURE_CLIENT_ID }}
+        tenant-id: ${{ secrets.AZURE_TENANT_ID }}
+        subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
```
Comment (martincostello) [inline .github/workflows/build.yml:177]: These secret names are just copy-pasted from the sample - happy to change them to something else that's more obvious what they're for.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -131,28 +138,106 @@ jobs:
...
+    - name: Sign artifacts
+      shell: pwsh
+      run: >
+        ./sign code azure-key-vault
+        **/*.nupkg
+        --base-directory "${{ github.workspace }}/packages"
+        --file-list "${{ github.workspace }}/signing-config/filelist.txt"
+        --publisher-name "App vNext"
+        --description "Polly"
+        --description-url "https://github.com/App-vNext/Polly"
```
Comment (martincostello) [inline .github/workflows/build.yml:188]: Happy to change any/all of these to something different.

**Comment by martincostello** [comment]
Looks like this is working 🎉

I'm just working on an extra bit of the workflow to validate this non-manually.

image

image

**Comment by martincostello** [comment]
Validation is working too: [logs](https://github.com/App-vNext/Polly/actions/runs/5223663899/jobs/9431067035?pr=1269#step:6:89)

**Review by joelhulen — APPROVED** [review]

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -131,28 +138,159 @@ jobs:
...
+        if ($invalidPackages -gt 0) {
+          Write-Output "::error::$invalidPackages NuGet package(s) failed validation."
+        }
+
+    - name: Checkout vcsjones/AuthenticodeLint
+      uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
+      with:
+        path: AuthenticodeLint
+        ref: ccfaec53ee5c1b14f029cb8156e0653c530f8b65
+        repository: vcsjones/AuthenticodeLint
```
[the remark this answers is not in the document]
Comment (martincostello) [reply .github/workflows/build.yml:226]: I tried both `dotnet verify` and `NuGetKeyVaultSignTool verify` as a way to remove the need to compile this tool to do the validation, but neither work correctly for Authenticode signatures at this time unfortunately: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/issues/1219#issuecomment-1584908168

**Pull request merged**